src/main/java/com/commercetools/sync/sdk2/states/utils/StateReferenceResolutionUtils.java

```diff
@@ -0,0 +1,107 @@
+import com.commercetools.api.models.state.StateResourceIdentifier;
+import com.commercetools.api.models.state.StateResourceIdentifierBuilder;
+import com.commercetools.sync.sdk2.commons.utils.ReferenceIdToKeyCache;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+
+/**
+ * Util class which provides utilities that can be used when syncing resources from a source
+ * commercetools project to a target one.
+ */
+public final class StateReferenceResolutionUtils {
+
+  /**
+   * Returns an {@link java.util.List}&lt;{@link StateDraft}&gt; consisting of the results of
+   * applying the mapping from {@link State} to {@link StateDraft} with considering reference
+   * resolution.
+   *
+   * <table>
+   *   <caption>Mapping of Reference fields for the reference resolution</caption>
+   *   <thead>
+   *     <tr>
+   *       <th>Reference field</th>
+   *       <th>from</th>
+   *       <th>to</th>
+   *     </tr>
+   *   </thead>
+   *   <tbody>
+   *     <tr>
+   *        <td>transitions</td>
+   *        <td>{@link java.util.List}&lt;{@link StateReference}&gt;</td>
+   *        <td>{@link java.util.List}&lt;{@link StateReference}&gt; (with key replaced with id field)</td>
+   *     </tr>
+   *   </tbody>
+   * </table>
+   *
+   * <p><b>Note:</b> The transition references should be cached(idToKey value fetched and stored in
+   * a map). Any reference, which have its id in place and not replaced by the key, it would not be
+   * found in the map. In this case, this reference will be considered as existing resources on the
+   * target commercetools project and the library will issues an update/create API request without
+   * reference resolution..
+   *
+   * @param states the states without expansion of references.
+   * @param referenceIdToKeyCache the instance that manages cache.
+   * @return a {@link java.util.List} of {@link StateDraft} built from the supplied {@link
+   *     java.util.List} of {@link State}.
+   */
+  @Nonnull
+  public static List<StateDraft> mapToStateDrafts(
+      @Nonnull final List<State> states,
+      @Nonnull final ReferenceIdToKeyCache referenceIdToKeyCache) {
+    return states.stream()
+        .filter(Objects::nonNull)
+        .map(
+            state -> {
+              final List<StateResourceIdentifier> newTransitions =
+                  replaceTransitionIdsWithKeys(state, referenceIdToKeyCache);
+              return StateDraftBuilder.of()
+                  .key(state.getKey())
+                  .type(state.getType())
+                  .name(state.getName())
+                  .description(state.getDescription())
+                  .initial(state.getInitial())
+                  .roles(state.getRoles())
+                  .transitions(newTransitions)
+                  .build();
+            })
+        .collect(Collectors.toList());
+  }
+
+  @SuppressWarnings("PMD.ReturnEmptyCollectionRatherThanNull")
```

Comment:
This violation makes sense. Return emptyList instead of null.

Reply:
It has different meanings when I return emptyList or null. Please check here: https://github.com/commercetools/commercetools-sdk-java-v2/blob/b1ad14b3323c125d1b38a1f1b29cf92e8f269055/commercetools/commercetools-sdk-java-api/src/main/java-generated/com/commercetools/api/models/state/StateDraftBuilder.java#L193